server/topic.go

```diff
@@ -930,60 +947,83 @@ func (t *Topic) handleBroadcast(msg *ServerComMessage) {
 			}
 
 			var read, recv, unread, seq int
-			if msg.Info.What == "read" {
-				if msg.Info.SeqId <= pud.readID {
-					// No need to report stale or bogus read status.
-					return
+			if asChan {
+				if msg.Info.What == "read" {
+					read = msg.Info.SeqId
+					seq = read
+				} else if msg.Info.What == "recv" {
+					recv = msg.Info.SeqId
+					seq = recv
 				}
+			} else {
+				if msg.Info.What == "read" {
+					if msg.Info.SeqId <= pud.readID {
+						// No need to report stale or bogus read status.
+						return
+					}
 
-				// The number of unread messages has decreased, negative value.
-				unread = pud.readID - msg.Info.SeqId
-				pud.readID = msg.Info.SeqId
-				if pud.readID > pud.recvID {
-					pud.recvID = pud.readID
-				}
-				read = pud.readID
-				seq = read
-			} else if msg.Info.What == "recv" {
-				if msg.Info.SeqId <= pud.recvID {
-					// Stale or bogus recv status.
-					return
-				}
+					// The number of unread messages has decreased, negative value.
+					unread = pud.readID - msg.Info.SeqId
+					pud.readID = msg.Info.SeqId
+					if pud.readID > pud.recvID {
+						pud.recvID = pud.readID
+					}
+					read = pud.readID
+					seq = read
+				} else if msg.Info.What == "recv" {
+					if msg.Info.SeqId <= pud.recvID {
+						// Stale or bogus recv status.
+						return
+					}
 
-				pud.recvID = msg.Info.SeqId
-				if pud.readID > pud.recvID {
-					pud.recvID = pud.readID
+					pud.recvID = msg.Info.SeqId
+					if pud.readID > pud.recvID {
+						pud.recvID = pud.readID
+					}
+					recv = pud.recvID
+					seq = recv
 				}
-				recv = pud.recvID
-				seq = recv
 			}
 
 			if seq > 0 && !t.isProxy {
-				if err := store.Subs.Update(t.name, asUser,
-					map[string]interface{}{
-						"RecvSeqId": pud.recvID,
-						"ReadSeqId": pud.readID},
-					false); err != nil {
+				topicName := t.name
+				if asChan {
+					topicName = msg.Info.Topic
+				}
 
+				upd := map[string]interface{}{}
+				if recv > 0 {
+					upd["RecvSeqId"] = recv
+				}
+				if read > 0 {
+					upd["ReadSeqId"] = read
+				}
+				if err := store.Subs.Update(topicName, asUid, upd, false); err != nil {
```

Comment:
(Perhaps that won't happen often if at all) but do we need to update subs here if `upd` is empty?

Reply:
It cannot be empty: `seq > 0` means wither read or recv or both are > 0.